pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SingleSnapshotAbortedTxnProcessorImpl.java

```diff
@@ -87,7 +90,16 @@ public CompletableFuture<PositionImpl> recoverFromSnapshot() {
                     PositionImpl startReadCursorPosition = null;
                     try {
                         while (reader.hasMoreEvents()) {
-                            Message<TransactionBufferSnapshot> message = reader.readNext();
+                            Message<TransactionBufferSnapshot> message = reader.readNext(2, TimeUnit.SECONDS);
```

Comment:
It treats a message that cannot be received in 2 seconds as an exceptional case that the message has been deleted. I think it should be done in `SystemTopicClient#readNext` instead of adding another public API. If there are other places (e.g. a protocol handler that could access this interface) that calls `SystemTopicClient#readNext` without the timeout, the error should be processed in the same way.

Reply:
I'd rather implement the wrapped `readNext()` like:

```java
        public Message<T> readNext() throws PulsarClientException {
            Message<T> message = reader.readNext(2, TimeUnit.SECONDS);
            if (message == null) {
                throw new PulsarClientException.InvalidMessageException(
                        "When reading from topic " + reader.getTopic() + ", the latest message has been deleted");
            }
            return message;
        }
```

Then the exception could be handled in the catch case:

```java
                    try {
                        while (reader.hasMoreEvents()) {
                            Message<TransactionBufferSnapshot> message = reader.readNext();
                            /* ... */
                    } catch (PulsarClientException.InvalidMessageException e) {
                        return FutureUtil.failedFuture(
                                new TransactionBufferException.TBRecoverCantCompletedException(e.getMessage()));
```

Reply:
In addition, I'm wondering if the 2 seconds is reasonable. What if the message actually exists and the timeout happened only because of the network issue?

Reply:
>  I think it should be done in SystemTopicClient#readNext instead of adding another public API. If there are other places (e.g. a protocol handler that could access this interface) that calls SystemTopicClient#readNext without the timeout, the error should be processed in the same way.

I think it is a good suggestion. This also leads to a lack of flexibility in the API. For example, if there is no message at the moment and I need to wait for the following message to be sent, there is no API to use.  Since the 'SystemTopicClient.Reader' is an internal-only API, I feel it would be better to provide better usability.

> In addition, I'm wondering if the 2 seconds is reasonable. What if the message actually exists and the timeout happened only because of the network issue?

Yes, you are right. we can't identify what is happening.

The current fix is like a patch to fix this existing issue, and there will be one discussion to fix the root cause:
- When the client already gets the last message id and caches it, then the compaction task will delete the messages end of this topic, do change the logic of the compaction task: the last message of this topic will be retained

Reply:
Hi @BewareMyPower 

I used the simper logic suggested by @congbobo184, could you retake a look?